src/board_controller/brainalive/brainalive.cpp

```diff
@@ -371,21 +386,45 @@ void BrainAlive::adapter_1_on_scan_found (
 void BrainAlive::read_data (simpleble_uuid_t service, simpleble_uuid_t characteristic,
     uint8_t *data, size_t size, int channel_num)
 {
-    if (size != BRAINALIVE_PACKET_SIZE)
+    static int SOFTWARE_GAIN, HARDWARE_GAIN, REFFRENCE_VOLTAGE;
+    if (size == BRAINALIVE_HANDSHAKING_PACKET_SIZE)
+    {
+        // printf ("inside 6\n");
+        SOFTWARE_GAIN = data[1];
+        HARDWARE_GAIN = data[2];
+        REFFRENCE_VOLTAGE = (data[3] << 8) | data[4];
+    }
+    else if (size != BRAINALIVE_PACKET_SIZE)
     {
         safe_logger (spdlog::level::warn, "unknown size of BrainAlive Data {}", size);
         return;
     }
-
-    for (int i = 0; i < (int)size; i += 32)
+    else
     {
-        double eeg_data[9] = {0};
-        for (int j = i + 4, k = 0; j < i + 28; j += 3, k++)
+
+        for (int i = 0; i < (int)size; i += BRAINALIVE_SINGLE_PACKET_SIZE)
         {
-            eeg_data[k] = (((data[j] << 16 | data[j + 1] << 8 | data[j + 2]) << 8) >> 8) *
-                BRAINALIVE_EEG_SCALE_FACTOR / BRAINALIVE_EEG_GAIN_VALUE;
+            double ba_data[15] = {0};
```

Comment:
I dont understand how it works, here the size is 15 and hardcoded, in brainflow_boards.cpp num_rows is 18, and it should match each other, otherwise it will not be aligned in memory properly. I would recommend to allocate this buffer dynamically once somewhere using smth like this https://github.com/brainflow-dev/brainflow/blob/master/src/board_controller/enophone/enophone.cpp#L103 and release it in the end, This way you will avoid misalignments